docs/source/Mapping.rst

```diff
@@ -0,0 +1,152 @@
+        qc.measure_all()
+
+        print(qc.draw(fold=-1))
+
+    .. code-block:: console
+
+              ┌───┐                ░ ┌───┐ ░                 ░ ┌─┐
+         q_0: ┤ H ├──■────■────■───░─┤ T ├─░───■────■────■───░─┤M├─────────
+              └───┘┌─┴─┐  │    │   ░ ├───┤ ░   │    │  ┌─┴─┐ ░ └╥┘┌─┐
+         q_1: ─────┤ X ├──┼────┼───░─┤ T ├─░───┼────┼──┤ X ├─░──╫─┤M├──────
+                   └───┘┌─┴─┐  │   ░ ├───┤ ░   │  ┌─┴─┐└───┘ ░  ║ └╥┘┌─┐
+         q_2: ──────────┤ X ├──┼───░─┤ T ├─░───┼──┤ X ├──────░──╫──╫─┤M├───
+                        └───┘┌─┴─┐ ░ ├───┤ ░ ┌─┴─┐└───┘      ░  ║  ║ └╥┘┌─┐
+         q_3: ───────────────┤ X ├─░─┤ T ├─░─┤ X ├───────────░──╫──╫──╫─┤M├
+                             └───┘ ░ └───┘ ░ └───┘           ░  ║  ║  ║ └╥┘
+      meas: 4/══════════════════════════════════════════════════╩══╩══╩══╩═
+                                                          0  1  2  3
+
+
+Now assume this circuit shall be mapped to a 4-qubit architecture with the following coupling map:
+
+.. image:: /images/linear_arch_4-1.png
+   :width: 80%
+   :alt: 4-qubit Architecture
+   :align: center
+
+
+In *QMAP* this architecture can be manually defined.
+
+    .. code-block:: python3
+
+        arch = qmap.Architecture(4, set([(0, 1), (1, 2), (2, 3), (3, 2), (2, 1), (1, 0)]))
+
+The quantum circuit :code:`qc` can not be run directly on this architecture. Naive mapping yields the following possible compiled circuit.
+
+    .. code-block:: console
+
+                      ┌───┐                      ░ ┌───┐ ░                       ░ ┌─┐
+	  q_0 -> q_0: ┤ H ├──■───X───────────────░─┤ T ├─░───────────────X───■───░─┤M├─────────
+	              └───┘┌─┴─┐ │               ░ ├───┤ ░               │ ┌─┴─┐ ░ └╥┘┌─┐
+	  q_1 -> q_1: ─────┤ X ├─X───■───X───────░─┤ T ├─░───────X───■───X─┤ X ├─░──╫─┤M├──────
+	                   └───┘   ┌─┴─┐ │       ░ ├───┤ ░       │ ┌─┴─┐   └───┘ ░  ║ └╥┘┌─┐
+	  q_2 -> q_2: ─────────────┤ X ├─X───■───░─┤ T ├─░───■───X─┤ X ├─────────░──╫──╫─┤M├───
+	                           └───┘   ┌─┴─┐ ░ ├───┤ ░ ┌─┴─┐   └───┘         ░  ║  ║ └╥┘┌─┐
+	  q_3 -> q_3: ─────────────────────┤ X ├─░─┤ T ├─░─┤ X ├─────────────────░──╫──╫──╫─┤M├
+	                                   └───┘ ░ └───┘ ░ └───┘                 ░  ║  ║  ║ └╥┘
+	      meas: 4/══════════════════════════════════════════════════════════════╩══╩══╩══╩═
+                                                                                    0  1  2  3
+
+This mapping uses 4 SWAPs. As every SWAP decomposes into 3 CNOT gates, this is a very costly overhead for such a small circuit.
+Minimizing the number of additional gates introduced by mapping a quantum circuit is precisely the job of *QMAP*.
+
+Exact Mapping
+#############
+
+The *exact mapper* implemented in *QMAP* maps quantum circuits using the *minimal* number of SWAP gates. For directional architectures it furthermore tries to minimize the number of additional Hadamard gates. Using the exact mapper is as simple as:
```

Comment:
```suggestion
The *exact mapper* implemented in *QMAP* maps quantum circuits using the *minimal* number of SWAP gates.
To this end, it encodes the mapping task as a MaxSAT problem and subsequently solves it using the `SMT solver Z3 <https://github.com/Z3Prover/z3>`_. Due to the NP-hardness of the mapping task, this approach is only scalable up to roughly eight qubits in most scenarios.

.. note::
    On directional architectures, it can be significantly cheaper to surround a CNOT gate with four Hadamard operations (effectively exchanging its control and target) instead of adding a SWAP gate. For these architectures, QMAP minimizes the number of additional SWAP *and* H gates.

Using the exact mapper is as simple as:
```

Reply:
Done